src/app/common/UpgradeCheck.vue

```diff
@@ -42,6 +41,28 @@ const latestVersion = ref('')
 const showNotice = ref(false)
 
 checkVersion()
+// mostly taken from semver-compare
+const compare = (a: string, b: string) => {
+  const pa = a.split('.')
+  const pb = b.split('.')
+  for (let i = 0; i < 3; i++) {
+    const na = Number(pa[i])
+    const nb = Number(pb[i])
+    if (i === 0) {
+      // if 'major' isn't a number then prefer the alternate
+      if (isNaN(na)) {
+        return 1
+      } else if (isNaN(nb)) {
+        return 0
+      }
+    }
+    if (na > nb) return 1
+    if (nb > na) return -1
+    if (!isNaN(na) && isNaN(nb)) return 1
+    if (isNaN(na) && !isNaN(nb)) return -1
+  }
+  return 0
+}
 
 async function checkVersion(): Promise<void> {
   try {
```

Comment:
Yes definitely I just copied/pasted it from https://cdn.jsdelivr.net/npm/semver-compare@1.0.0/index.js 😅 and tweaked it slightly for what I needed. Let me take a better pass at this. I am right in thinking we never have to deal with ranges at all right (so `~` and `^`)

Reply:
Never we always exact versions

Reply:
K I changed this up slightly. The intention is:

- If the current version does not follow the exact pattern of `number.number.number` then don't even make the HTTP request to get the reported version from kuma.io. This means things like `dev` (or any other string), `1.1.1-preview` etc will **not** show the upgrade notice.
- Anything else goes through the comparison, during which time we are guaranteed to be dealing with numbers.

I also remove the `finally` and just returned early instead. Let me know what you think, especially around whether we should or should not show the upgrade notice when a string is involved.

---

Lastly, I noticed something very weird here:

https://github.com/kumahq/kuma-gui/blob/7b1a72afc3e94c56ce287f01dd1f480281e6ded6/src/app/common/UpgradeCheck.vue#L60-L66

Looking back at the original PR it looks like the intention was to save a date and wait 3 months to check again or something (not sure if I followed correctly)?:

> If the internet is NOT accessible from the GUI, the component will store today's date in localStorage (for accessing it later) and 3 months from that date, it will display the notice

Original PR: https://github.com/kumahq/kuma-gui/pull/31

Looks like the localStorage bits where removed before the PR got merged:

https://github.com/kumahq/kuma-gui/pull/31/commits/a8d7d1daee20142dd4e4d4f560f594f22fb2b350

Anyway, I'm not totally sure what the intention is here and I decided to stop looking at it and wondering whats happening, but maybe we should add an issue to figure out what we want and come back and revisit this? wdyt?

Reply:
Resolving this convo so I can merge with full 🟢 , we can chat later if there are any changes needed